File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/generic/GenericSmokerBlock.java
package generations.gg.generations.core.generationscore.common.world.level.block.generic;

import generations.gg.generations.core.generationscore.common.world.level.block.entities.GenerationsBlockEntities;
import generations.gg.generations.core.generationscore.common.world.level.block.entities.generic.GenericSmokerBlockEntity;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.SmokerBlock;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityTicker;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.BlockState;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class GenericSmokerBlock extends SmokerBlock {
    public GenericSmokerBlock() {
        super(BlockBehaviour.Properties.copy(Blocks.SMOKER));
    }

    @Override
    public BlockEntity newBlockEntity(@NotNull BlockPos pos, @NotNull BlockState state) {
        return new GenericSmokerBlockEntity(pos, state);
    }

    @Override
    @Nullable
    public <T extends BlockEntity> BlockEntityTicker<T> getTicker(@NotNull Level level, @NotNull BlockState state, @NotNull BlockEntityType<T> blockEntityType) {
        return SmokerBlock.createFurnaceTicker(level, blockEntityType, GenerationsBlockEntities.GENERIC_SMOKER.get());
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/world/level/block/generic/GenericSmokerBlock.kt
package generations.gg.generations.core.generationscore.common.world.level.block.generic

import com.mojang.serialization.MapCodec
import generations.gg.generations.core.generationscore.common.world.level.block.asValue
import generations.gg.generations.core.generationscore.common.world.level.block.entities.GenerationsBlockEntities
import generations.gg.generations.core.generationscore.common.world.level.block.entities.generic.GenericSmokerBlockEntity
import net.minecraft.core.BlockPos
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.SmokerBlock
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityTicker
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState

class GenericSmokerBlock(properties: Properties = Properties.ofFullCopy(Blocks.SMOKER)) : SmokerBlock(properties) {
    override fun codec(): MapCodec<SmokerBlock> = CODEC

    override fun newBlockEntity(pos: BlockPos, state: BlockState): BlockEntity? {
        return GenericSmokerBlockEntity(pos, state)
    }

    override fun <T : BlockEntity?> getTicker(
        level: Level,
        state: BlockState,
        blockEntityType: BlockEntityType<T>
    ): BlockEntityTicker<T>? {
        return createFurnaceTicker(level, blockEntityType, GenerationsBlockEntities.GENERIC_SMOKER.asValue())
    }

    companion object {
        val CODEC = simpleCodec<SmokerBlock>(::GenericSmokerBlock)
    }
}
